### ISSUE TITLE
Empty result is decoded as an empty string even when `output` is set to `json`

### ISSUE THREAD

**Issue by grenik** [sanack/node-jq#689]
If running a jq filter produces an empty result, the result is an empty string (`''`) even if the `output` option is set to `json`.

It may be a bit confusing because an empty string is also a different valid result (see examples below).

```js
// this filter yields no results
jq.run('select(.foo == "bar")', {foo: ""}, {input: "json", output: "json"}).then(result => {
  // result is an empty string ''
});

// this filter yields an empty string
jq.run('.foo', {foo: ""}, {input: "json", output: "json"}).then(result => {
  // result is an empty string ''
});
```

I believe that `undefined` would be a better option because it is used neither in JSON nor in JQ.

**Comment by davesnx** [comment]
Exactly, the option `output:"json"` is not guaranteed to be always JSON, it just tries to parse as JSON and in case of failing it keeps it as the stdout (which is a string). There's a few cases of queries where empty string is actually the result of the run.

In case of changing this behaviour to return `undefined` (only when the parsing fails) it can be miss-leading to assume that undefined could be returned as a operation that expects a JSON as output. Which arguably, is the same situation as the current approach, where we do the minimum to ensure the output is JSON.

I would like to change it, and either reject the promise or radically return an empty object. Any extra thoughts @mackermans?

Thanks for opening the issue @grenik, I believe isn't the first confusing case.

**Comment by mackermans** [comment]
Thanks for raising this, @grenik.

@davesnx I opened the PR thinking that `undefined` made the most sense. But now I understand your point 😄  We've made it pretty difficult for ourselves by returning parsed JSON, effectively making it a JS data type, and to make matters worse: falling back to raw JSON in case of failure.

I don't think the alternatives are better, though:
- Rejecting the promise indicates a parsing failure
- Returning an empty object breaks with JQ's behavior

Could we consider a new output mode (JS)?

**Comment by davesnx** [comment]
Indeed, the empty string is an expected result from jq saying that there's no filters/matches with your query. Either by outputting json or strings.

**Comment by davesnx** [comment]
I like the idea of the undefined, maybe we should have a breaking change and implement this. Are you up for it?
